clang/lib/Sema/SemaExpr.cpp

```diff
@@ -229,6 +229,20 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
                !checkAllowedSYCLInitializer(VD, /*CheckValueDependent =*/true))
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelConstStaticVariable;
+      else if (IsConst && VD->hasGlobalStorage()) {
```

Comment:
Is the limit only for const arrays?

Reply:
Per SYCL spec: 6.3 Language restrictions for kernels

> Variables with **static storage duration** that are odr-used inside a kernel must be const or constexpr and
> zero-initialized or constant-initialized.

Global variables (among others) have static storage duration. And SYCL allows using globals within a kernel as long as it is a const. This change is only for a very large constant array whose size exceeds SPIR-V's minimum limits.